torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,109 @@
+        # If no epsilon is added here, division by zero may be caused by vanishing gradients.
+        # This epsilon is not needed if the input matrix covers the gradients of at least one entire layer in the neural network.
+        if epsilon > 0:
+            col += epsilon
+        # Project it on the rest and remove it.
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def _set_random(vector, random_seed=0):
+    """
+    Initializes a 1D tensor from a standard normal distribution.
+    The see makes sure that the randomized vector in all the DDP replicas are the same at every step.
+    """
+    torch.manual_seed(np.random.RandomState(random_seed).randint(1_000_000_000))
+    vector.data[:] = torch.randn(*vector.shape)
+
+
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD graidient compression
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution;
+    2) Allreduces P;
+    3) Orthogonizes P;
+    4) Compute Q, which is approximately equal to M^TP;
+    5) Allreduces Q;
+    6) Computes M, which is approximately equal to PQ^T.
+    7) Truncates the input tensor to the original length.
+
+    Example::
+        PowerSGDState state(process_group, 1)
+        >>> ddp_model.register_comm_hook(state, powerSGD_hook)
+    """
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+    world_size = (
+        process_group.size() if process_group is not None else dist.get_world_size()
+    )
+
+    # The input tensor is a flattened 1D tensor.
+    input_tensor = bucket.get_tensors()[0]
+    device = input_tensor.device
+    total_length = input_tensor.shape[0]
+
+    # View the input tensor as a 2D square-shape tensor, and pad 0s if necessary.
+    square_side_length = math.ceil(math.sqrt(total_length))
+    padded_total_length = square_side_length ** 2
+    input_tensor.resize_(padded_total_length)
+    input_tensor[total_length:padded_total_length].fill_(0)
+    matrix = input_tensor.view(square_side_length, square_side_length)
+
+    def create_low_rank_tensor():
+        "Returns a low-rank 2D tensor and the allocated contiguous memory."
+        memory = torch.empty(square_side_length * rank, device=device)
+        return memory.view(square_side_length, rank), memory
+
+    p, p_memory = create_low_rank_tensor()
+    q, q_memory = create_low_rank_tensor()
+    # Initialize each Q from a standard normal distribution.
+    _set_random(q)
```

Comment:
For each iteration, this should use a different 'seed'. 
I think it would be better to also orthogonalize Q here.

Reply:
It seems that you implemented the algorithm without 'warm start'. Is that to simplify the initial tests?

Reply:
> For each iteration, this should use a different 'seed'.
Now I used the length of input tensor as the seed. Added a TODO here to read it from an external state provided by the constructor of the hook.

> I think it would be better to also orthogonalize Q here.
Added this step. Curious: it seems that you commented out it in https://github.com/epfml/powersgd/blob/master/gradient_reducers.py#L677.



> It seems that you implemented the algorithm without 'warm start'. Is that to simplify the initial tests?
Correct. This is just the first version. I plan to add more optimizations iteratively.